src/tango/ip/fd_netlink.c

```diff
@@ -0,0 +1,760 @@
+  X( NDA_MASTER             , __VA_ARGS__ ) \
+  X( NDA_LINK_NETNSID       , __VA_ARGS__ ) \
+  X( NDA_SRC_VNI            , __VA_ARGS__ ) \
+  X( NDA_PROTOCOL           , __VA_ARGS__ ) \
+  X( NDA_FDB_EXT_ATTRS      , __VA_ARGS__ )
+
+char const *
+fd_nda_type_to_label( uint nda_type ) {
+# define FD_NDA_MATCH( LABEL, ... ) \
+  if( nda_type == LABEL ) return #LABEL;
+  FD_NL_NDA_TYPE(FD_NDA_MATCH,)
+  return "NDA_UNKNOWN";
+# undef FD_NDA_MATCH
+}
+
+
+int
+fd_nl_create_socket( ) {
+  int fd = socket( AF_NETLINK, SOCK_RAW | SOCK_NONBLOCK, NETLINK_ROUTE );
+
+  if( fd < 0 ) {
+    FD_LOG_WARNING(( "Unable to create netlink socket. Error: %d %s", errno,
+          strerror( errno ) ));
+    return -1;
+  }
+
+  struct sockaddr_nl sa;
+  fd_memset( &sa, 0, sizeof(sa) );
+  sa.nl_family = AF_NETLINK;
+  if( bind( fd, (void*)&sa, sizeof(sa) ) < 0 ) {
+    FD_LOG_WARNING(( "Unable to bind netlink socket. Error: %d %s",
+          errno, strerror( errno ) ));
+    close( fd );
+    return -1;
+  }
+
+  return fd;
+}
+
+
+void
+fd_nl_close_socket( int fd ) {
+  if( fd >= 0 ) {
+    close( fd );
+  }
+}
+
+int
+fd_nl_init( fd_nl_t * nl, uint seq ) {
+  nl->seq = seq;
+  nl->fd  = fd_nl_create_socket();
+
+  /* returns 1 for failure, 0 for success */
+  return nl->fd < 0 ? 1 : 0;
+}
+
+void
+fd_nl_fini( fd_nl_t * nl ) {
+  fd_nl_close_socket( nl->fd );
+}
+
+
+long
+fd_nl_load_route_table( fd_nl_t *             nl,
+                        fd_nl_route_entry_t * route_table,
+                        ulong                 route_table_cap ) {
+  int fd = nl->fd;
+  if( fd < 0 ) return -1;
+
+  /* format the request */
+
+  /* Request struct */
+  struct {
+    struct nlmsghdr nlh;  /* Netlink header */
+    struct rtmsg    rtm;  /* Payload - route message */
+  } nl_request;
+
+  fd_memset( &nl_request, 0, sizeof( nl_request ) );
+
+  uint seq = nl->seq++;
+
+  nl_request.nlh.nlmsg_type  = RTM_GETROUTE;  /* We wish to get routes */
+#if 0
+  /* CAP_NET_ADMIN required for NLM_F_ATOMIC */
+  nl_request.nlh.nlmsg_flags = NLM_F_REQUEST | NLM_F_ROOT | NLM_F_ATOMIC;
+#else
+  nl_request.nlh.nlmsg_flags = NLM_F_REQUEST | NLM_F_ROOT;
+#endif
+  nl_request.nlh.nlmsg_len   = sizeof(nl_request);
+  nl_request.nlh.nlmsg_pid   = 0;
+  nl_request.nlh.nlmsg_seq   = seq;
+
+  /* request only IPv4 routes */
+  nl_request.rtm.rtm_family  = AF_INET;
+
+  long sent = send( fd, &nl_request, sizeof(nl_request), 0 );
+  if( sent == -1 ) {
+    FD_LOG_WARNING(( "Unable to make netlink request. Error: %d %s", errno, strerror( errno ) ));
+    return -1;
+  }
+
+  /* receiving */
+  long len;
+
+  /* set alignment such that the returned data is aligned */
+  uchar __attribute__(( aligned(16) )) ibuf[4096]   = {0};
+
+  /* Pointer to the messages head */
+  struct nlmsghdr *h = (struct nlmsghdr *)ibuf;
+
+  /* Try to read the message in case of busy or interrupted call */
+  while(1) {
+    do {
+      len = recv( fd, ibuf, sizeof(ibuf), 0 );
```

Comment:
As above, either assert that len == expected, or you need to repeatedly call receive until we get all the bytes we want.

Reply:
We can't know ahead of time how many bytes were going to get back. Examples on the internet don't have code for partial reads. In fact, many use SOCK_DGRAM instead of SOCK_RAW. The docs say these aren't treated differently.
It's not clear how the read code /should/ be.

Reply:
I think you need to read the first few bytes to get nlmsg_len, then keep polling with recv until you have read at most nlmsg_len.